vnext/Chakra/ChakraBytecodeStore.cpp

```diff
@@ -0,0 +1,130 @@
+  m_bytecodeDirectory = executorScratchDirectory / BytecodeScratchSubdirectory;
+}
+
+std::unique_ptr<JSBigString> ChakraBytecodeStore::tryObtainCachedBytecode(const std::string &scriptUrl) {
+  auto bytecodePath = bytecodePathFromScriptUrl(scriptUrl);
+  if (!bytecodePath) {
+    return nullptr;
+  }
+
+  auto bytecodeBodyStr = std::make_unique<FileMappingBigString>(bytecodePath->string(), sizeof(BytecodeFileHeader));
+  if (!bytecodeBodyStr->file_data()) {
+    return nullptr;
+  }
+
+  // file_data will go backwards to the beginning of the file instead of going to the specified offset
+  auto header = reinterpret_cast<BytecodeFileHeader *>(bytecodeBodyStr->file_data());
+
+  ChakraVersionInfo curChakraVersion;
+  if (!curChakraVersion.initialize() || header->chakraVersion != curChakraVersion) {
+    return nullptr;
+  }
+
+  folly::Optional<uint64_t> scriptChecksum = calculateScriptChecksum(scriptUrl);
+  if (!scriptChecksum || header->scriptChecksum != scriptChecksum) {
+    return nullptr;
+  }
+
+  return bytecodeBodyStr;
+}
+
+void ChakraBytecodeStore::persistBytecode(const std::string &scriptUrl, const JSBigString &bytecode) {
+  ChakraVersionInfo versionInfo;
+  if (!versionInfo.initialize()) {
+    return;
+  }
+
+  auto bytecodePath = bytecodePathFromScriptUrl(scriptUrl);
+  if (!bytecodePath) {
+    return;
+  }
+
+  folly::Optional<uint64_t> scriptChecksum = calculateScriptChecksum(scriptUrl);
+  if (!scriptChecksum) {
+    return;
+  }
+
+  // bytecode is a binary representation, so we need to pass in the "b" flag to fopen_s
+  FILE *bytecodeFilePtr;
+  if (_wfopen_s(&bytecodeFilePtr, bytecodePath->c_str(), L"wb")) {
+    return;
+  }
+  std::unique_ptr<FILE, decltype(&fclose)> bytecodeFilePtrWrapper(bytecodeFilePtr, fclose);
+
+  // Write a junk header at first in case we fail
+  BytecodeFileHeader junkHeader{};
+
+  if (fwrite(&junkHeader, sizeof(BytecodeFileHeader), 1, bytecodeFilePtrWrapper.get()) != sizeof(BytecodeFileHeader) ||
+      fwrite(bytecode.c_str(), bytecode.size(), 1, bytecodeFilePtrWrapper.get()) != bytecode.size() ||
+      fflush(bytecodeFilePtrWrapper.get())) {
+    return;
+  }
+
+  // Write the real header now
+  BytecodeFileHeader header{CurrentBytecodeFileFormatVersion, *scriptChecksum, versionInfo};
+
+  fseek(bytecodeFilePtrWrapper.get(), 0, SEEK_SET);
+  if (!fwrite(&header, sizeof(BytecodeFileHeader), 1, bytecodeFilePtrWrapper.get()) ||
+      fflush(bytecodeFilePtrWrapper.get())) {
+    return;
+  }
+}
+
+folly::Optional<std::filesystem::path> ChakraBytecodeStore::bytecodePathFromScriptUrl(const std::string &scriptUrl) {
+  std::filesystem::path filename = std::filesystem::path(scriptUrl).filename();
+  if (filename.empty()) {
+    return folly::none;
+  }
+
+  return m_bytecodeDirectory / filename.concat(BytecodeFileExtension);
+}
+
+folly::Optional<uint64_t> ChakraBytecodeStore::calculateScriptChecksum(const std::string &scriptUrl) {
```

Comment:
>const std::string &scriptUrl [](start = 71, length = 28)

This is super dumb on my part. We already have the script. Don't need to go reload it.